File: binarytreezigzaglevelordertraversal.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def zigzagLevelOrder(self, root):
        """
        :type root: TreeNode
        :rtype: List[List[int]]
        """
        self.res = []
        self.dfs(root, 0)
        #for i in range(len(self.res)):
        #    if i%2 == 1:
        #        self.res[i] = self.res[i][::-1]
        return self.res
    def dfs(self, root, depth):
        if not root:
            return
        if len(self.res) == depth:
            self.res.append([])
        if depth%2 != 0:
            self.res[depth].insert(0, root.val)
        else:
            self.res[depth].append(root.val)
        self.dfs(root.left, depth+1)
        self.dfs(root.right, depth+1)
            

File: cpp/binarytreezigzaglevelordertraversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        vector<vector<int>> res;
        if (!root) return res;
        queue<TreeNode*> que;
        que.push(root);
        int index = 0;
        while (!que.empty()) {
            int num = que.size();
            vector<int> tmp;
            for (int i = 0; i < num; i++) {
                auto target = que.front();
                que.pop();
                if (target->left) que.push(target->left);
                if (target->right) que.push(target->right);
                tmp.push_back(target->val);
            }
            if (index%2) reverse(tmp.begin(), tmp.end());
            res.push_back(tmp);
            index++;
        }
        return res;
    }
};
